src/riak_core_gossip.erl

```diff
@@ -367,73 +371,291 @@ remove_from_cluster(Ring, ExitingNode) ->
-            end
+        Qualifiers ->
+            %% Look at the current allocated vnode counts for each qualifying
+            %% node, and find all qualifying nodes with the lowest of these
+            %% counts
+            [{Q0, BestCnt}|Others] = lists:keysort(2, Qualifiers),
+            PreferredCandidates =
+                [{Q0, BestCnt}|
+                    lists:takewhile(fun({_, C}) -> C == BestCnt end, Others)],
+            
+            %% Final selection of a node as a destination for this partition,
+            %% The node Counts must be updated to reflect this allocation, and
+            %% the RingFun applied to actually queue the transfer
+            {Rand, Seed2} = rand:uniform_s(length(PreferredCandidates), Seed),
+            {Chosen, BestCnt} = lists:nth(Rand, PreferredCandidates),
+            UpdRing = RingFun(P, Chosen, Ring),
+            UpdCounts =
+                lists:keyreplace(Chosen, 1, Counts, {Chosen, BestCnt + 1}),
+            simple_transfer(Rest,
+                            {RingFun, TargetN, ExitingNode},
+                            UpdRing,
+                            {Seed2, [{P, Chosen}|Prev], UpdCounts})
     end;
-attempt_simple_transfer(Seed, Ring, [{_, N}|Rest], TargetN, Exit, Idx, Last) ->
-    %% just keep track of seeing this node
-    attempt_simple_transfer(Seed, Ring, Rest, TargetN, Exit, Idx+1,
-                            lists:keyreplace(N, 1, Last, {N, Idx}));
-attempt_simple_transfer(_, Ring, [], _, _, _, _) ->
+simple_transfer([{P, N}|Rest], Statics, Ring, {Seed, Prev, Counts}) ->
+    %% This is already allocated to a node other than the exiting node, so
+    %% simply transition to the Previous ring accumulator
+    simple_transfer(Rest, Statics, Ring, {Seed, [{P, N}|Prev], Counts});
+simple_transfer([], _Statics, Ring, _LoopAccs) ->
     {ok, Ring}.
+
+
+%% ===================================================================
+%% Unit tests
+%% ===================================================================
+
+-ifdef(TEST).
+
+test_ring_fun(P, N, R) ->
+    lists:keyreplace(P, 1, R, {P, N}).
+
+count_nodes(TestRing) ->
+    CountFun =
+        fun({_P, N}, Acc) ->
+            case lists:keyfind(N, 1, Acc) of
+                false ->
+                    lists:ukeysort(1, [{N, 1}|Acc]);
+                {N, C} ->
+                    lists:ukeysort(1, [{N, C + 1}|Acc])
+            end
+        end,
+    lists:foldl(CountFun, [], TestRing).
+
+simple_transfer_simple_test() ->
+    R0 = [{0, n5}, {1, n1}, {2, n2}, {3, n3},
+            {4, n4}, {5, n5}, {6, n3}, {7, n2}],
+    SomeTime = {1632,989499,279637},
+    FixedSeed = rand:seed(exrop, SomeTime),
+    {ok, R1} =
+        simple_transfer(R0,
+                        {fun test_ring_fun/3, 3, n4},
+                        R0,
+                        {FixedSeed,
+                            [],
+                            lists:keydelete(n4, 1, count_nodes(R0))}),
```

Comment:
The Counts are computed using Members, which does not include the `leaving` node, hence allowed to delete the leaving node here.